File: implementation/transaction/src/main/java/com/financetracker/transaction/logic/model/RecurringTransaction.java
package com.financetracker.transaction.logic.model;

import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

@Getter
@NoArgsConstructor
@Table(name = "recurring_transaction")
@Entity
public class RecurringTransaction extends Transaction {

    private LocalDate startDate;
    private Periodicity periodicity;

    @Embedded
    @AttributeOverride(name = "amount", column = @Column(name = "amount", precision = 27, scale = 6))
    private MonetaryAmount fixedAmount;

    @JoinColumn(name = "transaction_id", referencedColumnName = "id")
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private Set<TransactionRecord> transactionRecords = new HashSet<>();

    @Builder(builderMethodName = "with")
    public RecurringTransaction(final String id,
                                final String name,
                                final String description,
                                final Type type,
                                final Set<Label> labels,
                                final Transfer transfer,
                                final LocalDate startDate,
                                final Periodicity periodicity,
                                final MonetaryAmount fixedAmount,
                                final Set<TransactionRecord> transactionRecords) {
        super(id, name, description, type, labels, transfer);
        this.startDate = startDate;
        this.periodicity = periodicity;
        this.fixedAmount = fixedAmount;
        this.transactionRecords = transactionRecords;
    }
}

File: implementation/ui-financetracker/src/app/recurring-transaction/recurringTransaction.ts
import { MonetaryAmount } from "../bank-account/bankAccount";
import { Periodicity, Transfer, TransferStatus, Type } from "../transaction/transaction";

export class RecurringTransaction {
    id: string;
    name: string;
    description: string;
    type: Type;
    labels: string[];
    transfer: Transfer;
    startDate: string;
    periodicity: Periodicity;
    fixedAmount: MonetaryAmount;
    transactionRecords: TransactionRecord[];

    constructor(id: string, name: string, description: string, type: Type, labels: string[], transfer: Transfer, startDate: string, periodicity: Periodicity, fixedAmount: MonetaryAmount, transactionRecords: TransactionRecord[]) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.type = type;
        this.labels = labels;
        this.transfer = transfer;
        this.startDate = startDate;
        this.periodicity = periodicity;
        this.fixedAmount = fixedAmount;
        this.transactionRecords = transactionRecords;
    }
}

export class TransactionRecord {
    id: string;
    date: string;
    amount: MonetaryAmount;
    transferStatus: TransferStatus;

    constructor(id: string, date: string, amount: MonetaryAmount, transferStatus: TransferStatus) {
        this.id = id;
        this.date = date;
        this.amount = amount;
        this.transferStatus = transferStatus;
    }

}
